xds/internal/xdsclient/channel.go

```diff
@@ -0,0 +1,288 @@
+		return nil, errors.New("xdsChannel: resourceTypeGetter is nil")
+	case opts.eventHandler == nil:
+		return nil, errors.New("xdsChannel: eventHandler is nil")
+	}
+
+	xc := &xdsChannel{
+		transport:          opts.transport,
+		serverConfig:       opts.serverConfig,
+		bootstrapConfig:    opts.bootstrapConfig,
+		resourceTypeGetter: opts.resourceTypeGetter,
+		eventHandler:       opts.eventHandler,
+		closed:             grpcsync.NewEvent(),
+	}
+
+	l := grpclog.Component("xds")
+	logPrefix := opts.logPrefix + fmt.Sprintf("[xds-channel %p] ", xc)
+	xc.logger = igrpclog.NewPrefixLogger(l, logPrefix)
+
+	if opts.backoff == nil {
+		opts.backoff = backoff.DefaultExponential.Backoff
+	}
+	xc.ads = ads.NewStreamImpl(ads.StreamOpts{
+		Transport:          xc.transport,
+		EventHandler:       xc,
+		Backoff:            opts.backoff,
+		NodeProto:          xc.bootstrapConfig.Node(),
+		WatchExpiryTimeout: opts.watchExpiryTimeout,
+		LogPrefix:          logPrefix,
+	})
+	xc.lrs = lrs.NewStreamImpl(lrs.StreamOpts{
+		Transport: xc.transport,
+		Backoff:   opts.backoff,
+		NodeProto: xc.bootstrapConfig.Node(),
+		LogPrefix: logPrefix,
+	})
+	return xc, nil
+}
+
+// The xdsChannel represents a client channel to the management server, and is
+// responsible for managing the lifecycle of the ADS and LRS streams and
+// invoking callbacks on interested authorities for various stream events.
+type xdsChannel struct {
+	// The following fields are initialized at creation time and are read-only
+	// after that, and hence need not be guarded by a mutex.
+	transport          transport.Interface           // Takes ownership of this transport (used to make streaming calls).
+	ads                *ads.StreamImpl               // An ADS stream to the management server.
+	lrs                *lrs.StreamImpl               // An LRS stream to the management server.
+	serverConfig       *bootstrap.ServerConfig       // Configuration of the server to connect to.
+	bootstrapConfig    *bootstrap.Config             // Complete bootstrap configuration, used to decode resources.
+	resourceTypeGetter func(string) xdsresource.Type // Function to retrieve resource parsing functionality, based on resource type.
+	eventHandler       channelEventHandler           // Callbacks for ADS stream events.
+	logger             *igrpclog.PrefixLogger        // Logger to use for logging.
+	closed             *grpcsync.Event               // Fired when the channel is closed.
+}
+
+func (xc *xdsChannel) close() {
+	xc.closed.Fire()
+	xc.ads.Stop()
+	xc.lrs.Stop()
+	xc.transport.Close()
+	xc.logger.Infof("Shutdown")
+}
+
+// reportLoad returns a load.Store that can be used to report load to the LRS, and a
+// function that can be called to stop reporting load.
+func (xc *xdsChannel) reportLoad() (*load.Store, func()) {
+	if xc.closed.HasFired() {
+		if xc.logger.V(2) {
+			xc.logger.Infof("Attempt to start load reporting on closed channel")
+		}
+		return nil, func() {}
```

Comment:
In a lot of these operations on the xdsChannel you check if closed, then check logger level, then log something at info. I think you can make a helper checkIfClosed(string somethingToLog) bool and here

```
if checkIfClosed("Attempt to start loading reporting on closed channel") {
        return nil, func() {}
}
```

Reply:
The problem with doing that is that if we have a format string in the log message and we pass arguments, then those arguments will be evaluated when the `checkIfClosed` method is called. What this means is that the arguments will be evaluated even in the case where `checkIfClosed` figures that either the channel is not closed, or log verbosity is not `2` or more. So, we would be doing useless work. The whole point of using the `logger.V` check is to avoid evalutating those arguments when we know that we are not going to log (apart from actually emitting logs only when certain verbosity is configured).

Reply:
Ah ok I see sounds good.